### PR TITLE
[Test me!] Full auto firemode issues

### LINKED ISSUE #1944: automatic gun bugs
## Round ID:
null

## Testmerges:
null

## Reproduction:
-fire auto gun in stasis bed
it fires normally. Firing a normal gun does not. Tested using WT autorifle compared to shotguns
-spam click automatic gun
it will ignore fire rate. used same comparisons as above, with other slower automatic guns as a control as well. perhaps ad a clickdelay or some shit
-attempting to examine, or do any other non-fire click action,, including pointing, fires automatic guns
-player direction does not change while firing fullauto guns.

### PR DESCRIPTION
**Pull request by Kmc2000**
## About The Pull Request

Fixes some wackiness with full auto firemode breaking guns. Oh god. Why. The bugs keep on coming and they keep on coming...

fix https://github.com/BeeStation/BeeStation-Hornet/issues/1944

## Why It's Good For The Game

I was unable to replicate, so I cannot attest to whether that is fixed or not:
https://github.com/BeeStation/BeeStation-Hornet/issues/1986

## Changelog
:cl:
fix: Fixed inconsistencies with full auto weapons.
/:cl:

### CODE REVIEW

File: code/modules/projectiles/autofire.dm [hunk 1]
```diff
@@ -48,15 +49,18 @@ This system lets you spray and pray with guns when dragging the mouse.
...
+	if(!autofire_target && world.time >= next_autofire)
+		start_autofiring(object)
+
+/obj/item/gun/proc/start_autofiring(object)
 	set waitfor = FALSE //Asynchronous processing is required here due to the while loop. In other words. Don't hold up the client's clicking while we run a loop that can go on for ages (potentially!)
-	if(world.time < next_autofire)
-		autofire_target = null
-		return FALSE
-	next_autofire = world.time + 0.25 SECONDS //Get out of here with your stupid autoclicker...
+	next_autofire = world.time + 1 SECONDS
```
Comment (francinum) [inline code/modules/projectiles/autofire.dm:58]: Should be calculated from a variable on the gun.

**Comment by PowerfulBacon** [comment]
Tested this, if you spam click you can sometimes fire 2 shots at the same time.
If you click, then hold right away the gun does not full auto and only fires when you release the hold.

**Comment by zeskorion** [comment]
the shift click examine firing / user direction bug is still there


additionally, i found a new issue while testing: there is no Clumsiness check on fullauto guns

**Comment by Kmc2000** [comment]
> the shift click examine firing / user direction bug is still there
> 
> additionally, i found a new issue while testing: there is no Clumsiness check on fullauto guns

I don't understand how, there's literally a line of code to check for that

**Comment by zeskorion** [comment]
> > the shift click examine firing / user direction bug is still there
> > additionally, i found a new issue while testing: there is no Clumsiness check on fullauto guns
> 
> I don't understand how, there's literally a line of code to check for that

try it. my testing shows it still happening, unless i did it wrong

**Comment by zeskorion** [comment]
honestly, i dont even understand why the new autofire shit was necessary? automatic to 1 worked just fine, and my firerate pr fixed the issue with firerate

**Comment by Kmc2000** [comment]
> honestly, i dont even understand why the new autofire shit was necessary? automatic to 1 worked just fine, and my firerate pr fixed the issue with firerate

Because we use it for our full auto ARs and one of your guys asked for it upstream.
Aditionally, the old auto fire code would cause people to get kicked by the anti spam measures. We had this a lot.

**Review by PowerfulBacon — APPROVED** [review]
Fixes most of the issues, if more pop up in the future we can do another PR.
The only issue I found was because temp gun has a speed of 0.
I'll wait to see what Zesk, or Ike thinks first though

**Comment by PowerfulBacon** [comment]
Another issue that I recently discovered is being on help intent and spamming allows you to fire a lot faster than being on harm intent, not sure if this fixes that or not.
Use the CTF rifles (which have a fire rate of 2), go on harm intent and they fire at a normal speed, but spam clicking on help intent allows you to unload the entire mag in a few seconds.

**Comment by ike709** [comment]
@Kmc2000 ^

**Review by zeskorion — CHANGES_REQUESTED** [review]
above bugs still need fixing, as do the ones that make you unable to throw or examine with automatic guns

**Comment by Kmc2000** [comment]
The throw + examine bugs are addressed in this PR

File: code/modules/projectiles/autofire.dm [hunk 2]
```diff
@@ -48,15 +48,24 @@ This system lets you spray and pray with guns when dragging the mouse.
...
 	var/mob/user = src.loc
 	autofire_target = object //When we start firing, we start firing at whatever you clicked on initially. When the user drags their mouse, this shall change.
 	while(autofire_target)  //While will only run while we have a user (loc) that is a mob, and we are being actively held by this mob, they have a client (as to prevent disconnecting mid-fight causing you to perma-fire) and of course, if we passed the previous check about autofiring.
 		if(can_fire_at(autofire_target, user))
+			user.setDir(get_dir(user, object))
```
Comment (ivanmixo) [inline code/modules/projectiles/autofire.dm:68]: This doesn't work, try
```suggestion
			user.face_atom(autofire_target)
```

File: code/modules/projectiles/autofire.dm [hunk 3]
```diff
@@ -24,16 +24,16 @@ This system lets you spray and pray with guns when dragging the mouse.
...
 
 	if(isliving(user))//Check if the user can use the gun, if the user isn't alive(turrets) assume it can.
 		var/mob/living/L = user
 		if(!can_trigger_gun(L))
 			return FALSE
-		if(clumsy_check)//If theyre a clown, they can't autofire.
-			if (HAS_TRAIT(user, TRAIT_CLUMSY))
-				return FALSE
+	if(clumsy_check && HAS_TRAIT(user, TRAIT_CLUMSY))//If theyre a clown, they can't autofire.
+		return FALSE
```
Comment (ivanmixo) [inline code/modules/projectiles/autofire.dm:37]: I also suggest you add this so people can't shoot out of lockers or sleepers

```suggestion
if(!isturf(user.loc))
	return FALSE
```

**Comment by PowerfulBacon** [comment]
@Kmc2000 you keep saying these problems aren't on NSV which is weird but they totally are, I tried putting guns on racks there and ended up blasting the fuck out of them

**Comment by Kmc2000** [comment]
>bacon
You're the first person that I've heard who has a bug with them on nsv..

**Comment by PowerfulBacon** [comment]
> > bacon
> > You're the first person that I've heard who has a bug with them on nsv..

Wait really? Try putting a pulse rifle onto a rack and you just shoot, and you cant throw them either

**Comment by ike709** [comment]
Will this PR ever be done?

**Comment by PowerfulBacon** [comment]
Please can we just merge the changes, some critical bugs have been fixed that are still in game until we merge this

**Comment by ike709** [comment]
@PowerfulBacon Those two code suggestions are good suggestions though. KMC literally just needs to hit commit and fix it if github mangles the formatting.

**Comment by ivanmixo** [comment]
wew how has this still not been touched since I made those suggestions.

Almost tazed cap with hos gun I found while trying to throw it to him a couple rounds ago

**Comment by PowerfulBacon** [comment]
@Kmc2000 problems with full auto weapons shooting when you shift click, click on chargers or click on tables are on NSV too, valve plz fix

**Comment by Kmc2000** [comment]
Gimmie a hot sec.

**Pull request closed without merging**